README.md
# NAME

SQL::Translator::Parser::OpenAPI - convert OpenAPI schema to SQL::Translator schema

# PROJECT STATUS

| OS      |  Build status |
|:-------:|--------------:|
| Linux   | [![Build Status](https://travis-ci.org/mohawk2/SQL-Translator-Parser-OpenAPI.svg?branch=master)](https://travis-ci.org/mohawk2/SQL-Translator-Parser-OpenAPI) |

[![CPAN version](https://badge.fury.io/pl/SQL-Translator-Parser-OpenAPI.svg)](https://metacpan.org/pod/SQL::Translator::Parser::OpenAPI) [![Coverage Status](https://coveralls.io/repos/github/mohawk2/SQL-Translator-Parser-OpenAPI/badge.svg?branch=master)](https://coveralls.io/github/mohawk2/SQL-Translator-Parser-OpenAPI?branch=master)

# SYNOPSIS

    use SQL::Translator;
    use SQL::Translator::Parser::OpenAPI;

    my $translator = SQL::Translator->new;
    $translator->parser("OpenAPI");
    $translator->producer("YAML");
    $translator->translate($file);

    # or...
    $ sqlt -f OpenAPI -t MySQL <my-openapi.json >my-mysqlschema.sql

    # or, applying an overlay:
    $ perl -MHash::Merge=merge -Mojo \
      -e 'print j merge map j(f($_)->slurp), @ARGV' \
        t/06-corpus.json t/06-corpus.json.overlay |
      sqlt -f OpenAPI -t MySQL >my-mysqlschema.sql

# DESCRIPTION

This module implements a [SQL::Translator::Parser](https://metacpan.org/pod/SQL::Translator::Parser) to convert
a [JSON::Validator::OpenAPI](https://metacpan.org/pod/JSON::Validator::OpenAPI) specification to a [SQL::Translator::Schema](https://metacpan.org/pod/SQL::Translator::Schema).

It uses, from the given API spec, the given "definitions" to generate
tables in an RDBMS with suitable columns and types.

To try to make the data model represent the "real" data, it applies heuristics:

- to remove object definitions considered non-fundamental; see
["definitions\_non\_fundamental"](#definitions_non_fundamental).
- for definitions that have `allOf`, either merge them together if there
is a `discriminator`, or absorb properties from referred definitions
- creates object definitions for any properties that are an object
- creates object definitions for any properties that are an array of simple
OpenAPI types (e.g. `string`)
- creates object definitions for any objects that are
`additionalProperties` (i.e. freeform key/value pairs), that are
key/value rows
- absorbs any definitions that are in fact not objects, into the referring
property
- injects foreign-key relationships for array-of-object properties, and
creates many-to-many tables for any two-way array relationships

# ARGUMENTS

## snake\_case

If true, will create table names that are not the definition names, but
instead the pluralised snake\_case version, in line with SQL convention. By
default, the tables will be named after simply the definitions.

# PACKAGE FUNCTIONS

## parse

Standard as per [SQL::Translator::Parser](https://metacpan.org/pod/SQL::Translator::Parser). The input $data is a scalar
that can be understood as a [JSON::Validator
specification](https://metacpan.org/pod/JSON::Validator#schema).

## defs2mask

Given a hashref that is a JSON pointer to an OpenAPI spec's
`/definitions`, returns a hashref that maps each definition name to a
bitmask. The bitmask is set from each property name in that definition,
according to its order in the complete sorted list of all property names
in the definitions. Not exported. E.g.

    # properties:
    my $defs = {
      d1 => {
        properties => {
          p1 => 'string',
          p2 => 'string',
        },
      },
      d2 => {
        properties => {
          p2 => 'string',
          p3 => 'string',
        },
      },
    };
    my $mask = SQL::Translator::Parser::OpenAPI::defs2mask($defs);
    # all prop names, sorted: qw(p1 p2 p3)
    # $mask:
    {
      d1 => (1 << 0) | (1 << 1),
      d2 => (1 << 1) | (1 << 2),
    }

## definitions\_non\_fundamental

Given the `definitions` of an OpenAPI spec, will return a hash-ref
mapping names of definitions considered non-fundamental to a
value. The value is either the name of another definition that _is_
fundamental, or or `undef` if it just contains e.g. a string. It will
instead be a reference to such a value if it is to an array of such.

This may be used e.g. to determine the "real" input or output of an
OpenAPI operation.

Non-fundamental is determined according to these heuristics:

- object definitions that only have one property (which the author calls
"thin objects"), or that have two properties, one of whose names has
the substring "count" (case-insensitive).
- object definitions that have all the same properties as another, and
are not the shortest-named one between the two.
- object definitions whose properties are a strict subset of another.

# OPENAPI SPEC EXTENSIONS

## `x-id-field`

Under `/definitions/$defname`, a key of `x-id-field` will name a
field within the `properties` to be the unique ID for that entity.
If it is not given, the `id` field will be used if in the spec, or
created if not.

This will form the ostensible "key" for the generated table. If the
key used here is an integer type, it will also be the primary key,
being a suitable "natural" key. If not, then a "surrogate" key (with a
generated name starting with `_relational_id`) will be added as the primary
key. If a surrogate key is made, the natural key will be given a unique
constraint and index, making it still suitable for lookups. Foreign key
relations will however be constructed using the relational primary key,
be that surrogate if created, or natural.

## `x-view-of`

Under `/definitions/$defname`, a key of `x-view-of` will name another
definition (NB: not a full JSON pointer). That will make `$defname`
not be created as a table. The handling of creating the "view" of the
relevant table is left to the CRUD implementation. This gives it scope
to use things like the current requesting user, or web parameters,
which otherwise would require a parameterised view. These are not widely
available.

## `x-artifact`

Under `/definitions/$defname/properties/$propname`, a key of
`x-artifact` with a true value will indicate this is not to be stored,
and will not cause a column to be created. The value will instead be
derived by other means. The value of this key may become the definition
of that derivation.

## `x-input-only`

Under `/definitions/$defname/properties/$propname`, a key of
`x-input-only` with a true value will indicate this is not to be stored,
and will not cause a column to be created. This may end up being merged
with `x-artifact`.

# DEBUGGING

To debug, set environment variable `SQLTP_OPENAPI_DEBUG` to a true value.

# AUTHOR

Ed J, `<etj at cpan.org>`

# LICENSE

Copyright (C) Ed J

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO

[SQL::Translator](https://metacpan.org/pod/SQL::Translator).

[SQL::Translator::Parser](https://metacpan.org/pod/SQL::Translator::Parser).

[JSON::Validator::OpenAPI](https://metacpan.org/pod/JSON::Validator::OpenAPI).


Changes
- require JSON::Validator 4.13 as has J:V:Ref with CLEAR method

0.07	Tue  5 Feb 05:27:26 GMT 2019
 - update to dep on JSON::Validator::OpenAPI::Mojolicious as actually use that

0.06	Tue  5 Feb 05:13:26 GMT 2019
 - doc tweaks
 - add x-id-field
 - snake_case arg
 - update to JSON::Validator 3 - fix #2 (thanks @rabbiveesh for report)

0.05	Mon 29 Oct 03:48:22 GMT 2018
 - thin object also if 2 properties, 1 =~ /count/i
 - add x-artifact, x-input-only, x-view-of
 - implement self-many-to-many

0.04	Sat 15 Sep 22:39:33 BST 2018
 - absorb non-object definitions
 - many-to-many tables now have composite PK, not separate "id", and
   columns named so DCSL names rels right
 - update docs

0.03	Wed 12 Sep 08:31:43 BST 2018
 - PK is already unique
 - use Math::BigInt to mean unlimited no of properties
 - use MySQL as dump instead of YAML
 - merge together allOf properties noting "discriminator" as superset
 - handle enum
 - no zap referenced definitions
 - extract free-standing "object" into new "definition"
 - turn additionalProperties objects into array of key/value pairs
 - field comments
 - extract array simple types
 - create many-to-many tables where mutual array-rels

0.02	Tue  4 Sep 17:51:10 BST 2018
 - update metadata for kwalitee (thanks @manwar!)
 - update metadata to dodge slight data-structure diffs

0.01	Mon  3 Sep 11:51:55 BST 2018
 - first version


t/lib/JTTest.pm
package JTTest;

my @IMPORT;
BEGIN {
@IMPORT = qw(
  strict
  warnings
  Test::More
  Test::Snapshot
  SQL::Translator
  SQL::Translator::Parser::OpenAPI
);
do { eval "use $_; 1" or die $@ } for @IMPORT;
}

use parent 'Exporter';
use Import::Into;

# nothing yet
our @EXPORT = qw(
  run_test
);

sub import {
  my $class = shift;
  my $target = caller;
  $class->export_to_level(1);
  $_->import::into(1) for @IMPORT;
}

sub run_test {
  my ($snake_case) = @_;
  (my $file = $0) =~ s#schema\.t$#corpus.json#;
  $file =~ s#json$#yml# if !-f $file;
  die "$file: $!" if !-f $file;

  require JSON::Validator::OpenAPI; # loads JSON and YAML loaders
  my $openapi_schema = JSON::Validator::OpenAPI->new->schema($file)->schema->data;

  my $overlay = "$file.overlay";
  if (-f $overlay) {
    my $data = do { open my $fh, $overlay or die "$overlay: $!"; local $/; <$fh> };
    my $overlay_data = $file =~ /json$/
      ? Mojo::JSON::decode_json($data)
      : YAML::XS::Load($data);
    require Hash::Merge;
    $openapi_schema = Hash::Merge::merge($openapi_schema, $overlay_data);
  }

  my $translator = SQL::Translator->new;
  $translator->parser("OpenAPI");
  $translator->parser_args(snake_case => $snake_case);
  $translator->producer("SQLite");

  my $got = $translator->translate(data => $openapi_schema);
  if ($got) {
    my @lines = split /\n/, $got;
    splice @lines, 0, 4; # zap opening blurb to dodge false negs
    $got = join "\n", @lines;
  } else {
    diag $translator->error;
  }
  is_deeply_snapshot $got, ($snake_case ? 'schema' : 'schema_camel');
}

1;


Makefile.PL
use 5.008001;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
  NAME             => 'SQL::Translator::Parser::OpenAPI',
  AUTHOR           => q{Ed J <etj@cpan.org>},
  VERSION_FROM     => 'lib/SQL/Translator/Parser/OpenAPI.pm',
  ABSTRACT_FROM    => 'lib/SQL/Translator/Parser/OpenAPI.pm',
  LICENSE          => 'artistic_2',
  MIN_PERL_VERSION => '5.014',
  CONFIGURE_REQUIRES => {
    'ExtUtils::MakeMaker' => '7.10',
  },
  TEST_REQUIRES => {
    'Test::More' => '0.98',
    'Test::Snapshot' => '0.02', # better snapshot filenames
    'YAML::XS' => '0.67', # YAML loader used by JV
  },
  PREREQ_PM => {
    'SQL::Translator' => '0.11024', # 0.11021 gives YAML diffs on "comments"
    'JSON::Validator::OpenAPI::Mojolicious' => '0', # after pointless zapping of ::OpenAPI from J::V
    'String::CamelCase' => '0',
    'SQL::Translator::Schema::Constants' => 0,
    'Lingua::EN::Inflect::Number' => '0',
    'Math::BigInt' => '0',
    'Hash::MoreUtils' => '0.06',
    'Hash::Merge' => '0',
  },
  clean => { FILES => 'SQL-Translator-Parser-OpenAPI-*' },
  META_MERGE => {
    "meta-spec" => { version => 2 },
    dynamic_config => 0,
    resources => {
      x_IRC => 'irc://irc.perl.org/#graphql-perl',
      repository => {
        type => 'git',
        url => 'git@github.com:mohawk2/SQL-Translator-Parser-OpenAPI.git',
        web => 'https://github.com/mohawk2/SQL-Translator-Parser-OpenAPI',
      },
      bugtracker  => {
        web => 'https://github.com/mohawk2/SQL-Translator-Parser-OpenAPI/issues',
      },
      license => [ 'http://dev.perl.org/licenses/' ],
    },
    prereqs => {
      develop => {
        requires => {
          'Test::Pod' => '1.22',
          'Pod::Markdown' => 0,
        },
      },
    },
  },
);

sub MY::postamble {
  <<EOF;
pure_all :: README.md

README.md : \$(VERSION_FROM)
\tpod2markdown \$< >\$\@
EOF
}
